### PR TITLE
Use ubuntu 16

### PR DESCRIPTION
**Pull request by wangkuiyi**
Fixes https://github.com/PaddlePaddle/Paddle/issues/2039

I tested this PR by

1. run `docker build -t paddle:dev .` to build the development Docker image.
1. run `docker run ... paddle:dev` to build the source code of PaddlePaddle.
1. run `cd build; docker build -t paddle .` to build the production image.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -51,11 +53,12 @@ RUN pip install --upgrade pip && \
 RUN apt-get install -y libssl-dev libffi-dev
 RUN pip install certifi urllib3[secure]
 
-RUN curl -sSL https://cmake.org/files/v3.4/cmake-3.4.1.tar.gz | tar -xz && \
-    cd cmake-3.4.1 && ./bootstrap && make -j `nproc` && make install && \
-    cd .. && rm -rf cmake-3.4.1
-
-VOLUME ["/woboq_out"]
+# Install woboq_codebrowser to /woboq
```
Comment (gangliao) [inline Dockerfile]: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/scripts/docker/build.sh#L67
改这里

File: paddle/api/CMakeLists.txt [hunk 2]
```diff
@@ -92,7 +92,9 @@ add_custom_command(OUTPUT ${PROJ_ROOT}/paddle/py_paddle/_swig_paddle.so
 # TODO(yuyang18) : make wheel name calculated by cmake
 add_custom_target(python_api_wheel ALL DEPENDS ${PROJ_ROOT}/paddle/py_paddle/_swig_paddle.so)
 
-install(DIRECTORY ${PROJ_ROOT}/paddle/dist/ DESTINATION opt/paddle/share/wheels)
+install(DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/dist/
```
Comment (gangliao) [inline paddle/api/CMakeLists.txt]: This wheel package will be generated in `${PROJ_ROOT}/paddle/dist/` automatically. It cannot be controlled by CMake.

After compilation, I checked where is `*.whl`.
gangl@GangLiao ~/b/Paddle> find . -name "*.whl"
`./paddle/dist/py_paddle-0.10.0rc4-py2-none-any.whl`

`install(DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/dist/ DESTINATION opt/paddle/share/wheels)`
 means that copy `*.whl` from `${CMAKE_CURRENT_BINARY_DIR}/dist/` to `DESTINATION opt/paddle/share/wheels`

Because `*.whl` does not exist in `${CMAKE_CURRENT_BINARY_DIR}/dist/`, it will cause the compile error...

Comment (wangkuiyi) [inline paddle/api/CMakeLists.txt]: I see. Does that mean that we should just keep it `${PROJ_ROOT}/paddle/dist`?

File: paddle/scripts/docker/build.sh [hunk 3]
```diff
@@ -83,11 +108,25 @@ if [[ ${WOBOQ:-OFF} == 'ON' ]]; then
...
+# generate deb package for current build
+# FIXME(typhoonzero): should we remove paddle/scripts/deb ?
+# FIXME: CPACK_DEBIAN_PACKAGE_DEPENDS removes all dev dependencies, must
+# install them in docker
+cat <<EOF
+========================================
+Generating .deb package ...
+========================================
+EOF
+cpack -D CPACK_GENERATOR='DEB' -D CPACK_DEBIAN_PACKAGE_DEPENDS="" ..
```
Comment (gangliao) [inline paddle/scripts/docker/build.sh]: @typhoonzero

File: doc/howto/dev/write_docs_en.rst [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+Document files of PaddlePaddle are in sub-directory :code:`doc`.  Source files are in `RST <http://docutils.sourceforge.net/docs/user/rst/quickref.html>`_ format.  We can build the document by letting `cmake<https://cmake.org/>`_ invoke `sphinx <http://www.sphinx-doc.org/en/1.4.8/>`_  to convert RST files into HTML files.
+
+
+How to Build Documents
+======================
+
+To save the time of installing building tools, we provide a Docker image.
+
+
+使用Docker构建PaddlePaddle的文档
```
Comment (gangliao) [inline doc/howto/dev/write_docs_en.rst]: All written work in`write_docs_en.rst` must be in English.

**Comment by gangliao** [comment]
Please resolve conflicting files

File: paddle/scripts/docker/build.sh (hunk 3 above)
Comment (typhoonzero) [inline paddle/scripts/docker/build.sh]: Should remove `-D CPACK_DEBIAN_PACKAGE_DEPENDS=""`.  I added this before because the base image is `python:2.7.13-slim` which installs python by building from source code. Install Paddle `wheel` package will install another version of python in the image, so I removed the dependencies. Now we are using `apt-get` to install python and python header files, so dependencies will not break when install.

File: paddle/api/CMakeLists.txt (hunk 2 above)
Comment (gangliao) [inline paddle/api/CMakeLists.txt]: Yeah. ${CMAKE_CURRENT_SOURCE_DIR} means the current source code directory.
${CMAKE_CURRENT_SOURCE_DIR} denotes the root directory of source code.

File: doc/howto/dev/write_docs_en.rst (hunk 4 above)
Comment (wangkuiyi) [inline doc/howto/dev/write_docs_en.rst]: Let me remove this file from this PR and do it in another PR.

File: paddle/scripts/docker/build.sh (hunk 3 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/build.sh]: Got it and removed!

File: doc/howto/dev/write_docs_cn.rst [hunk 5]
```diff
@@ -2,14 +2,13 @@
 如何贡献/修改文档
 ##################
 
-PaddlePaddle的文档包括英文文档 ``doc`` 和中文文档 ``doc_cn`` 两个部分。文档都是通过 `cmake`_ 驱动 `sphinx`_ 编译生成，生成后的文档分别存储在编译目录的 ``doc`` 和 ``doc_cn`` 两个子目录下。
+PaddlePaddle的文档文件都在 :code:`doc` 这个子目录里。源文件是 `RST <http://docutils.sourceforge.net/docs/user/rst/quickref.html>`_ 格式的。 在编译PaddlePaddle源码的时候，可以选择让 `cmake<https://cmake.org/>`_ 调用 `sphinx <http://www.sphinx-doc.org/en/1.4.8/>`_ 从 RST 文件生成 HTML 格式的文档。
```
Comment (typhoonzero) [inline doc/howto/dev/write_docs_cn.rst]: 不知道为什么但 `cmake<https://cmake.org/>`_  这里的链接好像没有正确生成。

Comment (wangkuiyi) [inline doc/howto/dev/write_docs_cn.rst]: @typhoonzero You are right. I am removing the claim of fixing #2041 from this PR. Also, I am abandoning my change to the Chinese document so to make this PR focusing on "Using Ubuntu 16".
